Update Platform Checkout Tracking conditions

malithsen opened the pull request:
#### Changes proposed in this Pull Request

With the recent changes in the usage tracking( p2: p2y3YZ-6L8-p2) , we can now collect usage data specific to WooPay regardless of the usage tracking opt-in status. This PR does the following so that we can receive `platform_checkout_enabled` and `platform_checkout_disabled` events regardless of usage tracking opt-in status.

- A minor refactor to change where additional event properties are being recorded.
- Introduce `maybe_record_admin_event` to record events emitted from the admin area
- Use `maybe_record_admin_event` to capture `platform_checkout_enabled` and `platform_checkout_disabled` events.
<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Regression Test
- Make sure WooPay is enabled.
- Add a product to the cart and go to the checkout page.
- In ~5 minutes, check the Tracks Live View for the `woocommerceanalytics_order_checkout_start` event.
- Disable WooPay
- Add a product to the cart and go to the checkout page.
- `woocommerceanalytics_order_checkout_start` event should not appear in the Tracks Live view.

Check Admin events

- Optin to usage tracking in WooCommerce -> Settings -> Advanced -> WooCommerce.com -> Enable Tracking.
- Do the tests mentioned in this PR https://github.com/Automattic/woocommerce-payments/pull/3816. Note that setting `_wcpay_feature_platform_checkout` flag is no longer needed.
- Opt-out of tracking by disabling the checkbox set in step 1.
- Repeat the tests mentioned in https://github.com/Automattic/woocommerce-payments/pull/3816
- `wcadmin_platform_checkout_enabled` and `wcadmin_platform_checkout_disabled` events should still be registered.

Optional
- For good measure smoke test checkout with split upe enabled/disabled

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

malithsen committed:
Change where the additional Tracks props are being recorded

malithsen committed:
Implement WooPay admin event tracking

Instead of using wc_admin_record_tracks_events for WooPay specific
events, we now record events via Platform Checkout Tracker. This allows
more granular tracking of Platform Checkout specific events when is is enabled.

malithsen committed:
Rename variable

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6175` or branch name `update/platform-checkout-admin-tracks` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=update/platform-checkout-admin-tracks) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 2c638f3a87a6b20f35ac75c13b839b192a51a587
- Build time: 2023-05-12 15:30:02 UTC

_Note: the build is updated when a new commit is pushed to this PR._

malithsen committed:
Fix split-upe constructor

malithsen committed:
Add changelog

malithsen committed:
Instantiate platform tracker closer to where it's used

malithsen committed:
Refactor Tracker instantiation to simplify reuse

malithsen committed:
Make unshared function private

This was made public in a previous commit, but that approach was
abandoned.

malithsen commented:
Hi @pierorocca 
@leonardola noted the following while reviewing the changes in this PR to update `platform_checkout_enabled` and `platform_checkout_disabled` events so that they are recorded regardless of the merchant's Usage Tracking opt-in status.

```
>  If you disable WooCommerce Usage Tracking we will not collect data from your site through our analytics tool, but we may still collect similar data if we need it to operate a service we provide to you, such as WooCommerce Payments.

To me that means no analytics but we will collect just enough data via other tools like WC Pay Server to make payments possible.
```

I know we've previously discussed that these two events would fall under product improvement data, but I could not find a good source to share with Leo to confirm it. Could you please confirm that we are okay to record these events for all WCPay merchants with the updated policy.

pierorocca commented:
@malithsen the main source is Neil's writeup in this P2 https://automattlock.wordpress.com/2023/04/18/making-our-wctracker-usage-tracking-page-more-accurate/

And to double check, I asked him to confirm the events in the spreadsheet against the usage tracking terms:
![image](https://github.com/Automattic/woocommerce-payments/assets/109928953/76f779d1-7397-4d04-8d1d-965a6034bc45)

leonardola commented on includes/class-platform-checkout-tracker.php:
```diff
@@ -154,20 +165,22 @@ public function should_enable_tracking() {
 		}
 
-		if ( ! $this->should_enable_tracking() ) {
+		if ( ! $is_admin_event && ! $this->should_enable_tracking() ) {
```
shouldn't this check actually happen in the `maybe_` functions?

malithsen replied:
`is_admin_event` identifies the events coming through the `maybe_record_admin_event`. This check makes sure that those requests bail out early without considering the conditions in `should_enable_tracking` method, as we want these events to be tracked always. 
But after you pointed it out, I realized it would make sense to consolidate all the validations in `should_enable_tracking`, so I moved this check inside that function in https://github.com/Automattic/woocommerce-payments/pull/6175/commits/36da0f3ddfe323c5e54646fd76c90f634d02fdee

leonardola replied:
What I meant is that it sounds strange to have checks in the function with a name that makes it look like it will execute something no matter what and have no checks in the functions with a name that indicate it may or may not execute. It's just a nitpicking and can be ignored.

leonardola commented on includes/class-platform-checkout-tracker.php:
```diff
@@ -197,9 +210,18 @@ public function tracks_record_event( $user, $event_name, $properties = [] ) {
 
+		// Add event property for test mode vs. live mode events.
+		$properties['test_mode']     = WC_Payments::mode()->is_test() ? 1 : 0;
+		$properties['wcpay_version'] = get_option( 'woocommerce_woocommerce_payments_version' );
```
Shouldn't we actually use the `WCPAY_VERSION_NUMBER` const instead of `get_option` as it will not change during downgrades and it will be calling the DB every time?

malithsen replied:
Good point. Thanks! I updated it in https://github.com/Automattic/woocommerce-payments/pull/6175/commits/390939be3d952a255ffbf960a38535dc7ebf2d4e

leonardola reviewed (commented):
Code generally looks good but when I added a debugger breakpoint to [this line](https://github.com/Automattic/woocommerce-payments/blob/update%2Fplatform-checkout-admin-tracks/includes/class-platform-checkout-tracker.php#L176) and visited the checkout page as a shopper I noticed `woocommerceanalytics_platform_checkout_express_button_offered` is also being recorded.

Also I left a couple of questions below

malithsen committed:
Consolidate validations

- Move tracking validations into the should_enable_tracking function
- Add more comments to clarify validations

malithsen committed:
Get WCPay version from the constant instead of the db

malithsen committed:
Fix WooPay Tracker class path

malithsen commented:
Thanks for the review @leonardola! I addressed your code comments above.

> Code generally looks good but when I added a debugger breakpoint to [this line](https://github.com/Automattic/woocommerce-payments/blob/update%2Fplatform-checkout-admin-tracks/includes/class-platform-checkout-tracker.php#L176) and visited the checkout page as a shopper I noticed woocommerceanalytics_platform_checkout_express_button_offered is also being recorded.

I misunderstood this when you first pointed it out on Slack. If you have WooPay enabled, what you see is the correct behavior. All Tracks events originating due to user behavior work the same way as before. That is, they are tracked when WooPay is enabled. 

I made some notes about existing tracking conditions [here](https://woopayp2.wordpress.com/woopay-data/) if it helps. Note that "Merchant Events" section should be updated following this PR.

Let me know if you have other questions.

malithsen committed:
Naming changes in comments

leonardola commented on includes/class-woopay-tracker.php:
```diff
@@ -98,33 +105,54 @@ public function ajax_tracks() {
+		// 3. Site page tracking is only enabled when WooPay is active.
+
 		// Track only site pages.
 		if ( is_admin() && ! wp_doing_ajax() ) {
```
We can probably remove this check now

malithsen replied:
This check is still needed. When an admin user goes through WooPay checkout, resulting events should not be tracked.

leonardola commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1770,7 +1771,7 @@ public function update_option( $key, $value = '' ) {
 		if ( $is_woopay_enabled !== $current_is_woopay_enabled ) {
-			wc_admin_record_tracks_event(
+			WC_Payments::woopay_tracker()->maybe_record_admin_event(
 				$is_woopay_enabled ? 'woopay_enabled' : 'woopay_disabled',
```
Just confirming that the events are now `wcadmin_woopay_enabled` and `wcadmin_woopay_disabled` and not `platform_checkout_enabled` and `platform_checkout_disabled`

malithsen replied:
Yes. This is correct. These events are renamed as a part of recently merged naming convention updates https://github.com/Automattic/woocommerce-payments/pull/6061

leonardola reviewed (approved):
It looks good to me. Just added a comment about a check that does not seem necessary anymore and about the event name

malithsen commented:
> What I meant is that it sounds strange to have checks in the function with a name that makes it look like it will execute something no matter what and have no checks in the functions with a name that indicate it may or may not execute. It's just a nitpicking and can be ignored.

Ah I see what you mean now. `maybe_` prefix was added to the top-level functions to indicate that they may not always result in actually tracking the event due to the validations that get triggered further down. 

I'll merge this as is but I acknowledge that it can be refactored to make it more clear.

pierorocca commented:
Nice work all!
